### DISCUSSION TITLE
JS-Friendly Snowflake Alternative

### DISCUSSION THREAD

**Discussion by inxilpro** [hirethunk/verbs#42] [Ideas]
I posted this in the [Verbs discord](https://discord.gg/hDhZmD6ZC9), but I'm going to re-post here for more visibility…

An issue with using Snowflakes everywhere is that javascript doesn't handle 64-bit integers very well (only the first 53 bits are accurately stored). That means that if you try to use a snowflake as an integer in JS, things may break. The solution is to treat snowflakes as strings, which is how the default `jsonSerialize` behavior of [`glhd/bits`](https://github.com/glhd/bits) works, but that doesn't help people who are echoing the IDs directly to their page…

I have a potential solution… but it's pretty technical.

Snowflakes are made up of 6 parts:

- 1 unused bit
- 41 bits of millisecond-accurate timestamp
- 5 bits for the datacenter ID
- 5 bits for the worker ID
- 12 bits for the internal sequence

The 12 bits of sequence lets you generate 4096 snowflakes per millisecond, per worker, per datacenter. That's a lot of parallel capacity. Arguably more than anyone other Twitter (the creator of the snowflake format) would need…

We could, theoretically, get a 53-bit version with some trade-offs. Traditional 32-bit signed unix timestamps have about +68 years of capacity (that's why we're hitting the year 2038 problem — 1970 + 68). But we don't need a sign bit, since snowflake IDs only move forward.

That means that if we accept second-accuracy, rather than millisecond-accuracy, we can fit about 136 years into 32 bits. So if we do the following:

- 32 bits for timestamp
- 4 bits for datacenter
- 5 bits for worker
- 12 bits for sequence

That would let you generate 4096 unique snowflakes per second, per worker, per datacenter, and produce a 53-bit value that could be represented in JS.

We'd need to pick a set of defaults that would cover most decent-scale apps (I'm not sure that 32/4/5/12 is right… but it's a decent example).

Any thoughts?

**Comment by inxilpro** [comment]
So Matt King found a few existing approaches to this same problem:

- https://github.com/cablehead/python-fity3
- https://samifayoumi.ca/blog/001_53bit-snowflakeid/

The `fity-3` approach is:

- 41-bit millisecond timestamp (~70 years)
- 8-bit worker ID (up to 256 parallel workers)
- 4-bit sequence (up to 16 ids, per worker, per millisecond)

The Sami Fayoumi approach is similar, trading a larger sequence for fewer workers:

- 41-bit timestamp
- 7-bit machine ID (128 workers)
- 5-bit sequence (32 ids, per worker, per millisecond)

Both approaches seem pretty good, the former allowing for more parallelization (which might be better suited to Laravel Vapor) while the latter allows for more individual throughput (you could generate 32 unique IDs inside a process before having to `usleep` for a single millisecond).

I definitely think that jumping from milliseconds to seconds is maybe too much, but I wonder if we could trim a few bits from the timestamp if we used `10ms` intervals, like [sonyflakes](https://github.com/sony/sonyflake)… In that case, we could eek out 109 years from 35 bits, and get ourselves a little more room for parallelization and throughput…

**Comment by ryanshipswell** [comment]
FWIW I've been using this https://github.com/kra8/laravel-snowflake for generating snowflake primary keys in apps. It supports a shorter 53bit snowflake version. I'm not sure what 11 bits it's cutting though, actually I think they are droping the unused but plus the data center and worker bits do maybe that's a bad example because you lose parallelization support.

Personally I think for most uses it makes sense to prioritize a longer timestamp and longer sequence over parallelization. At least for me personally I figure it'll be 2038 long before I need hundreds of servers.

Although I wonder if the answer isn't to just use a ULID instead of a snowflake.

**Reply by inxilpro** [reply]
If you're going to use a ULID, just treat the snowflake as a string (which is probably the right solution in general). You get all the benefits of Snowflakes in your DB and in PHP, and then you just treat them as strings in JS.

I think we may end up refactoring Verbs to use string versions of Snowflakes everywhere to be clear that they should be used as strings. Not sure, though.

**Reply by ryanshipswell** [reply]
Thinking about it some more, I like where you were going with the 35-bit 10ms interval, you could do something like this.

35-bit timestamp (10ms interval, 109 years)
9-bit machine/worker ID (512 workers)
9-bit sequence (512 ids, per worker, per 10ms)

It would take massive scale to need more than 512 workers IMO. I suspect most apps won't scale beyond tens of workers. So you could almost trade off less workers for more sequence if you wanted.

**Comment by peterfox** [comment]
I've had a similar problem before with CockroachDB and Livewire. CockroachDB uses random IDs instead of autoincrementing so they become large numbers that Javascript doesn't like. Surely the best thing is to make the IDs cast to a string via the model casts and in the event you need to use them for some kind of functionality you just grab the ID and convert it to an int?

**Comment by madebycaliper** [comment]
Not sure if this is relevant but it looks like Laravel just pushed support for unique string primary keys. PR here:

https://github.com/laravel/framework/pull/53280

**Comment by tsterker** [comment]
See also the [Verbs Discord](https://discord.gg/hDhZmD6ZC9) for related discussions:

- [Summary about the Snowflake vs. UUID/ULID Discussion](https://discord.com/channels/1176581731168567366/1176581732074532908/1394391485599973609)
- [Considerations about actual problem statement and concerns about accidentally shifting some of the pains if not properly planned/explained](https://discord.com/channels/1176581731168567366/1176581732074532908/1394459313375805606)
  - Summarized argument in [Verbs Discord](https://discord.com/channels/1176581731168567366/1176581732074532908/1399188551777648700)

**Comment by tsterker** [comment]
# ℹ️ Deterministic Snowflake -> ULID Migration

The crux of this comment is the **generation a deterministic ULID from a given Snowflake**. This would be relevant in most migration scenarios that move from Snowflake to ULID (or UUID v4).

I'm currently not looking into moving from Snowflake to ULID, but rather introduce a stable ULID alongside my Snowflakes as more robust (and explicit) identifier to use for the public (e.g. APIs and GUI). I still decided to provide my full (work-in-progress) migration path for more context. If you don't care about all the fluff, simply check out the **`UlidHelper`** below.

## Disclaimers
- The below is a bit rough around the edges, but I wanted to start sharing some of my progress for insight and potential scrutiny.

- My actual current use-case is to **_not_ move verbs away from Snowflake**, but instead introducing a dedicated `public_id` ULID that I use for the frontend  (related [discord thread](https://discord.com/channels/1176581731168567366/1176581732074532908/1394459313375805606)). Maybe my learnings could be useful in the context migration planning.

## Analysis & Planning

**Problem statement:**
- I have existing eloquent models that use a Snowflake `id` column
- I want to now introduce a new `public_id` ULID column that I'll use for public facing IDs (e.g. GUI and API routing keys)
- The new `public_id` column must be stable/deterministic across replays

**The approach:**
- I have existing past events that don't generate a `public_id`
  - I use this new `public_id` as new routing key (`getRoutingKeyName`)
- I'll update my event to deterministically derive the `public_id` from the Snowflake ID
- To not break my API that would now encounter only `null` values for `public_id`, I need to make sure to populate these by either:
  - A) Deriving the `public_id` during my migration
  - B) Migrate and immediately do a replay

## Implementation

### Migration update

First, I create a new migration that adds a nullable `public_id` column.
I need to add the `unique` constraint with a separate `ALTER TABLE` step to not run into integrity constraint issues.

```php
// 2025_07_27_add_public_id_to_my_models.php
return new class extends Migration
{
    public function up(): void
    {
        Schema::table('my_models', function (Blueprint $table) {
            $table->ulid('public_id')->nullable()->after('id');
        });

        // Add unique constraint after column creation and population
        Schema::table('my_models', function (Blueprint $table) {
            $table->unique('public_id');
        });
    }
};
```

### Update Model Routing Key

```diff
class MyModel extends Model
{
    use HasSnowflakes;

    protected $fillable = [
        'id',
+       'public_id',
    ];

    protected $casts = [
+       'public_id' => 'string',
    ];

+   public function getRouteKeyName(): string
+   {
+       return 'public_id';
+   }
}

```

### Change any "Frontent" code to use new `public_id`

Make sure that all relevant parts of your code now utilize the new `public_id`, instead of the internal `id` Snowflake column.

e.g.

```diff
           <a href="{{ route('my-models.show', [
-              'myModel' => $myModel->id
+              'myModel' => $myModel->public_id
            ]) }}">View My Model</a>
```

### Update Event that creates `MyModel`

This is where the rubber hits the road. We are **generating a deterministic ULID from the Verbs Snowflake**.

```diff
class MyEvent extends Event
{
    #[StateId(MyState::class)]
    public ?int $my_model_id = null;

    public function handle(MyState $state)
    {
        MyModel::create([
            'id' => $this->my_model_id,
+           'public_id' => UlidHelper::fromSnowflake($this->my_model_id),
        ]);
    }
}
```

## The `UlidHelper` to generate deterministic ULID from Snowflakes

> **EDIT(2025-07-27)**: I realized that my use case does not require the preservation of the snowflake timestamp information: https://github.com/hirethunk/verbs/discussions/42#discussioncomment-13903797

⚠️ Mostly **AI-generated**:

```php
<?php

namespace App\Support;

use Glhd\Bits\Snowflake;
use Symfony\Component\Uid\Ulid;

class UlidHelper
{
    /**
     * Generate a deterministic ULID from a 64-bit Snowflake ID
     */
    public static function fromSnowflake(int|string|Snowflake $snowflakeId): string
    {
        // Ensure we have an integer
        $snowflakeId = (int) (string) $snowflakeId;

        // glhd/bits style Snowflake: timestamp in top 42 bits, epoch starts 2023-01-01
        // @see https://github.com/glhd/bits/blob/f5ce0667323822cd1d6eb5a4aef699001c4c751f/config/bits.php#L32
        $epoch = 1672531200000;
        $timestampMs = ($snowflakeId >> 22) + $epoch;

        // --- Step 1: Convert timestamp to 48-bit binary string ---
        $timestampBin = str_pad(decbin($timestampMs), 48, '0', STR_PAD_LEFT);

        // --- Step 2: Use lower 22 bits + machine ID + sequence to generate entropy ---
        // We'll hash the full snowflake ID deterministically, then truncate to 80 bits
        $hash = hash('sha1', (string) $snowflakeId, true); // 160 bits (20 bytes)
        $entropyBin = '';
        for ($i = 0; strlen($entropyBin) < 80; $i++) {
            $entropyBin .= str_pad(decbin(ord($hash[$i])), 8, '0', STR_PAD_LEFT);
        }
        $entropyBin = substr($entropyBin, 0, 80); // Truncate to 80 bits

        // --- Step 3: Combine timestamp + entropy ---
        $ulidBin = $timestampBin . $entropyBin;

        // --- Step 4: Convert binary to bytes ---
        $ulidBytes = '';
        for ($i = 0; $i < 128; $i += 8) {
            $ulidBytes .= chr(bindec(substr($ulidBin, $i, 8)));
        }

        // --- Step 5: Encode as ULID string ---
        return (string) Ulid::fromBinary($ulidBytes);
    }
}
```

## Migration Path

To actually move into production with the above changes, we need to make sure that all our existing models will also receive a `public_id`.

Conceptually, this is roughly what needs to happen
- Run the DB migration to introduce the `public_id` column
- Update existing models to have a snowflake-derived `public_id`
- Activate the code that uses the `public_id` column
- Ensure there are no models missed (e.g. due to new events during your migration)

The key mechanisms are the updating of the `public_id`, which can happen either via (ignoring potential race conditions and assuming the app in maintenance mode for simplicity):

- A) Replay
- B) On-the-fly update during migration
- C) Combination of the above to minimize downtime

### A) Replay

Once you deployed the changes and ran your migration to introduce the new `public_id` column, you need to replay your events by:
- Truncating the relevant model tables
- Replay events

The nice thing about this is the fact that we truly event-source our changes and ensure it is working as expected.

### B) On-the-fly update during migrations

The on-the-fly update could utilize our `UlidHelper` to update existing models with a snowflaked-derived ID as part of our migration:

```diff
return new class extends Migration
{
    /**
     * Run the migrations.
     */
    public function up(): void
    {
        Schema::table('my_models', function (Blueprint $table) {
            $table->ulid('public_id')->nullable()->after('id');
        });

+       // Populate existing records with derived ULIDs from their snowflake IDs
+       MyModel::whereNull('public_id')->chunkById(100, function ($models) {
+           foreach ($models as $model) {
+               $model->update([
+                   'public_id' => UlidHelper::fromSnowflake($model->id),
+               ]);
+           }
+       });

        // Add unique constraint after column creation and population
        Schema::table('my_models', function (Blueprint $table) {
            $table->unique('public_id');
        });
    }
};

```

FYI @inxilpro @DanielCoulbourne

**Reply by tsterker** [reply]
**Update:** I realized that my use case (introducing a static ULID alongside the actual event Snowflake ID) would not require me to preserve the timestamp information of the original Snowflake ID. I only care about keeping my new ULID stable across replays. So, this could simplify my particular implementation.

But [as this might still be relevant requirement for the actual move of Verbs to UUID](https://discord.com/channels/1176581731168567366/1176581732074532908/1365382224979038331), my example `UlidHelper could still serve as a relevant example.
